File: inst/htmlwidgets/utils/events.js
/**
 * Subscribes to table events and attaches event callbacks.
 *
 * @param {Object} x - the object containing events to subscribe to
 * @param {string} tableId - the ID of the table
 * @param {Tabulator} tabulatorTable - the Tabulator table to subscribe events to
 */
const subscribeTableEvents = (x, tableId, tabulatorTable) => {
  if (!!x.events) {
    x.events.forEach(eventName => {
      const eventCallback = getEventCallback(tableId, eventName);
      tabulatorTable.on(eventName, eventCallback);
    })
  }
};

/**
 * Subscribes to default events of a Tabulator table and sets input values in Shiny.
 *
 * @param {Object} tabulatorTable - The Tabulator table instance.
 */
const subscribeDefaultEvents = (tabulatorTable) => {
  tabulatorTable.on("cellEditing", function(cell) {
    const cellData = {
      field: cell.getField(),
      initial_value: cell.getInitialValue(),
      old_value: cell.getOldValue(),
      new_value: cell.getValue(),
      row_position: cell.getRow().getPosition(),
    }
    Shiny.setInputValue(`${tabulatorTable.element.id}_editing`, cellData, { priority: "event" });
  });

  tabulatorTable.on("cellEdited", function(cell) {
    const cellData = {
      field: cell.getField(),
      initial_value: cell.getInitialValue(),
      old_value: cell.getOldValue(),
      new_value: cell.getValue(),
      row_position: cell.getRow().getPosition(),
    }
    Shiny.setInputValue(`${tabulatorTable.element.id}_edited`, cellData, { priority: "event" });
  });

  // After rendering, bind all of the shiny inputs associated to this instance,
  // if any.
  tabulatorTable.on("tableBuilt", function() {
    Shiny.bindAll(document.getElementById(tabulatorTable.element.id));
  });

};

// Column events that can be subscribed
const columnEventsTypes = {
  eventColumnEvents : [
    "headerClick", "headerDblClick", "headerContext", "headerTap", "headerDblTap",
    "headerTapHold", "headerMouseEnter", "headerMouseLeave", "headerMouseOver",
    "headerMouseOut", "headerMouseMove", "headerMouseDown", "headerMouseUp",
  ],
  columnColumnEvents : ["columnMoved",],
  columnEvents : ["columnResized", "columnTitleChanged",],
  columnVisibleEvents : ["columnVisibilityChanged",],
}

const rowEventsTypes = {
  eventRowEvents: [
    "rowClick", "rowDblClick", "rowContext", "rowTap", "rowDblTap", "rowTapHold",
    "rowMouseEnter", "rowMouseLeave", "rowMouseOver", "rowMouseOut", "rowMouseMove",
    "rowMouseDown", "rowMouseUp",
  ],
  rowEvents: ["rowAdded", "rowUpdated", "rowDeleted", "rowResized", ]
}

const spreadsheetEventsTypes = {
  eventSpreadsheetEvents: [
    "sheetAdded", "sheetRemoved", "sheetLoaded", "sheetUpdated",
  ]
}

const cellEventsTypes = {
  eventCellEvents: [
    "cellClick", "cellDblClick", "cellContext", "cellTap", "cellDblTap", "cellTapHold",
    "cellMouseEnter", "cellMouseLeave", "cellMouseOver", "cellMouseOut", "cellMouseMove",
    "cellMouseDown", "cellMouseUp",
  ]
}

const validationEventsTypes = {
  eventCellEvents: ["validationFailed", ],
};

const spreadsheetEventCallbacks = {
  sheetCallback: (tableId, eventName) => {
    return function(sheet) {
      const inputId = `${tableId}_${eventName}`;
      window.lasheet = sheet;
      const dataEvent = {
        event: eventName,
        sheet_key: sheet.getKey(),
        sheet_name: sheet.getTitle()
      }
      Shiny.setInputValue(inputId, dataEvent, { priority: "event" });
    }
  }
}

// Column callbacks that are passed to each event
const columnEventCallbacks = {
  eventColumnCallback : (tableId, eventName) => {
    return function(e, column) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          field: column.getField(),
          col_values: column.getCells().map(x => x.getValue())
        },
        { priority: "event" }
      );
    }
  },
  columnColumnCallback: (tableId, eventName) => {
    return function(column, columns) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          field: column.getField(),
          col_values: column.getCells().map(x => x.getValue()),
          cols_order: columns.map(x => x.getField())
        },
        { priority: "event" }
      );
    }
  },
  columnCallback: (tableId, eventName) => {
    return function(column) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          field: column.getField(),
          col_values: column.getCells().map(x => x.getValue())
        },
        { priority: "event" }
      );
    }
  },
  columnVisibleCallback: (tableId, eventName) => {
    return function(column, visible) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          field: column.getField(),
          col_values: column.getCells().map(x => x.getValue()),
          visible: visible
        },
        { priority: "event" }
      );
    }
  },
}

const rowEventCallbacks = {
  eventRowCallback: (tableId, eventName) => {
    return function(e, row) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          position: row.getPosition(),
          row_data: row.getData(),
        },
        { priority: "event" }
      );
    }
  },
  rowCallback: (tableId, eventName) => {
    return function(row) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          row_position: row.getPosition(),
          row_data: row.getData(),
        },
        { priority: "event" }
      );
    }
  },
}

const cellEventCallbacks = {
  cellEventCallback: (tableId, eventName) => {
    return function(e, cell) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          value: cell.getValue(),
          field: cell.getColumn().getField(),
          row_position: cell.getRow().getPosition()
        },
        { priority: "event" }
      );
    }
  }
}

const validationEventCallbacks = {
  /**
   * A function that handles a validation event callback.
   *
   * @param {string} tableId - the ID of the table
   * @param {string} eventName - the name of the event
   * @return {function} a function that takes cell, value, and validators as parameters
   */
  validationEventCallback: (tableId, eventName) => {
    return function(cell, value, validators) {
      const inputId = `${tableId}_${eventName}`;
      Shiny.setInputValue(
        inputId,
        {
          event: eventName,
          value: value,
          field: cell.getColumn().getField(),
          row_position: cell.getRow().getPosition(),
          validators: validators
        },
        { priority: "event" }
      );
    }
  }
}


/**
 * Get the event callback for a specific table and event name.
 *
 * @param {string} tableId - The ID of the table
 * @param {string} eventName - The name of the event
 * @return {any} The event callback function
 */
const getEventCallback  = (tableId, eventName) => {
  let callback;
  if (columnEventsTypes.eventColumnEvents.includes(eventName)) {
    callback = columnEventCallbacks.eventColumnCallback(tableId, eventName)
  } else if (columnEventsTypes.columnColumnEvents.includes(eventName)) {
    callback = columnEventCallbacks.columnColumnCallback(tableId, eventName)
  } else if (columnEventsTypes.columnEvents.includes(eventName)) {
    callback = columnEventCallbacks.columnCallback(tableId, eventName)
  } else if (columnEventsTypes.columnVisibleEvents.includes(eventName)) {
    callback = columnEventCallbacks.columnVisibleCallback(tableId, eventName)
  } else if (rowEventsTypes.eventRowEvents.includes(eventName)) {
    callback = rowEventCallbacks.eventRowCallback(tableId, eventName)
  } else if (rowEventsTypes.rowEvents.includes(eventName)) {
    callback = rowEventCallbacks.rowCallback(tableId, eventName)
  } else if (cellEventsTypes.eventCellEvents.includes(eventName)) {
    callback = cellEventCallbacks.cellEventCallback(tableId, eventName)
  } else if (spreadsheetEventsTypes.eventSpreadsheetEvents.includes(eventName)) {
    callback = spreadsheetEventCallbacks.sheetCallback(tableId, eventName)
  } else if (validationEventsTypes.eventCellEvents.includes(eventName)) {
    callback = validationEventCallbacks.validationEventCallback(tableId, eventName)
  }
  return callback
}

File: R/events.R
# API for interacting with JS events
# https://tabulator.info/docs/6.0/events#main-contents

event_names <- list(
  column_events = c(
    "headerClick", "headerDblClick", "headerContext", "headerTap", "headerDblTap",
    "headerTapHold", "headerMouseEnter", "headerMouseLeave", "headerMouseOver",
    "headerMouseOut", "headerMouseMove", "headerMouseDown", "headerMouseUp",
    "columnMoved", "columnResized", "columnTitleChanged", "columnVisibilityChanged"
  ),
  row_events = c(
    "rowClick", "rowDblClick", "rowContext", "rowTap", "rowDblTap", "rowTapHold",
    "rowMouseEnter", "rowMouseLeave", "rowMouseOver", "rowMouseOut", "rowMouseMove",
    "rowMouseDown", "rowMouseUp", "rowAdded", "rowUpdated", "rowDeleted", "rowResized"
  ),
  cell_events = c(
    "cellClick", "cellDblClick", "cellContext", "cellTap", "cellDblTap", "cellTapHold",
    "cellMouseEnter", "cellMouseLeave", "cellMouseOver", "cellMouseOut", "cellMouseMove",
    "cellMouseDown", "cellMouseUp"
  )
)


#' Checks for valid event names and returns a list of valid event names
#' @param event_name A list of event names to be checked
#' @return List of valid event names
check_valid_event_name <- function(event_name) {
  all_events <- unlist(event_names, use.names = FALSE)
  test <- all(event_name %in% all_events)
  if (!test) {
    #list events that are not valid
    invalid_events <- setdiff(event_name, all_events)
    cli_warn(
      c(
        "x" = "Invalid event names: {paste(invalid_events, collapse = ', ')}",
        "i" = "Check for the valid options using {.run tabulator::get_valid_event_names()}"
      )
    )
  }
  intersect(event_name, all_events)
}


#' Gets the valid event names to be passed to a tabulator object
#' @return List of valid event names
#' @export
get_valid_event_names <- function() {
  event_names
}

#' Subscribe events for the current table
#' Once events are subscribed, the events triggered in the browser can be listened
#' via registered inputs following the pattern `input$<table_id>_<eventName>`. For example
#' if the `cellClick` event was subscribed for a table with an id `my_table` then
#' you can use it in your Shiny app via the input registered to `input$table_cellClick`
#' Also you can run `run_example_app("table_events")` to see all the events available.
#' @param tabulator_object A tabulator object
#' @param events List of events to subscribe. Also can be a single character
#' @export
subscribe_events <- function(tabulator_object, events) {
  events_list <- if (length(events) == 1) list(events) else events
  events_list <- check_valid_event_name(events_list)
  if (length(events_list) > 0) {
    tabulator_object$x$events <- c(tabulator_object$x$events, events_list)
  }
  tabulator_object
}
